util/election/runner.go

```diff
@@ -0,0 +1,164 @@
+	// holds mastership.
+	MasterCheckInterval time.Duration
+	// MasterTTL is the time after which the instance will resign if encountered
+	// no successful IsMaster calls.
+	MasterTTL time.Duration
+}
+
+// Run contains mastership status of a single instance during a single run of
+// being master.
+type Run struct {
+	// Start is the time when the instance became the master as measured by the
+	// corresponding Runner's TimeSource.
+	Start time.Time
+	// Ctx is the context which is active while the instance believes to be the
+	// master. As soon as Ctx is canceled, the instance attempts to resign, and
+	// stops monitoring its status.
+	Ctx context.Context
+	// Cancel cancels Ctx, or does nothing if that is already canceled.
+	// Effectively, it instructs the instance to resign if not yet.
+	Cancel context.CancelFunc
+	// Done closes when mastership status monitoring goroutine for this run
+	// terminates, strictly after Ctx.Done() closes. After calling Cancel, a
+	// closed Done indicates completion of resigning (unless it failed, or Ctx
+	// parent context was canceled earlier).
+	Done <-chan struct{}
+}
+
+// Resign triggers mastership resignation, and returns when it is believed to
+// be completed.
+func (r *Run) Resign() {
+	r.Cancel()
+	<-r.Done
+}
+
+// Runner coordinates election participation for a single instance.
+//
+// An example usage pattern:
+//
+//   defer runner.Close(ctx)
+//   for { // We want to be the master potentially multiple times.
+//     run, err := runner.BeMaster(ctx)
+//     if err != nil {
+//       // Return if ctx canceled, or retry after some time.Sleep.
+//     }
+//     for { // Now we are the master.
+//       // Note: The work will be canceled if mastership is suddenly over.
+//       if err := DoSomeWorkAsMaster(run.Ctx); err != nil {
+//         if run.Ctx.Err() == err { // ctx or run.Ctx was canceled.
+//           break // Stop doing work as master.
+//         }
+//         // ... some other error, react accordingly ...
+//       }
+//       if ShouldResign(run.Start) {
+//         run.Resign()
+//         break
+//       }
+//     }
+//  }
+type Runner struct {
+	cfg        *Config
+	me         MasterElection
+	timeSource util.TimeSource // Allows for mocking tests.
+	prefix     string          // Prefix for all glog messages.
+}
+
+// NewRunner returns a new election runner for the provided configuration and
+// MasterElection interface.
+func NewRunner(cfg *Config, me MasterElection, ts util.TimeSource, prefix string) *Runner {
+	return &Runner{cfg: cfg, me: me, timeSource: ts, prefix: prefix}
+}
+
+// BeMaster blocks until it captures mastership on behalf of the current
+// instance, and returns a Run object which can be used to observe the
+// mastership status. Returns an error if capturing fails or the passed in
+// context is canceled before that.
+//
+// The method kicks off a goroutine which tracks mastership status, and
```

Comment:
The explanation of the goroutine's behaviour doesn't help much -- it's mostly just a translation of the code into text.  What's the *purpose* of the goroutine?  Is it "This method kicks off a goroutine that ensures mastership is always explicitly resigned" ?